caffe2/opt/optimize_ideep.cc

```diff
@@ -2,42 +2,404 @@
+#endif
+
 namespace caffe2 {
 namespace opt {
 
 using namespace nom;
 
-void OptimizeForIdeep(repr::NNModule* nn) {
-  // Conv+Relu fusion
-  auto should_fuse = [](const repr::Conv& conv) {
-    const auto annotation = conv.getAnnotation();
-    if (!annotation || !isa<Caffe2Annotation>(annotation)) {
-      return false;
+#ifndef CAFFE2_USE_IDEEP
+void OptimizeForIdeep(repr::NNModule* nn, caffe2::Workspace* ws, bool training_mode) {
+  LOG(WARNING) << "Only support optimizations for IDEEP";
+}
+
+#else
+USE_IDEEP_DEF_ALIASES();
+
+Blob* getBlob(repr::NNGraph::NodeRef node, caffe2::Workspace* ws) {
+  auto tensor = repr::nn::get<repr::Tensor>(node);
+  assert(ws->HasBlob(tensor->getName()) && "Blob not in workspace");
+  return ws->GetBlob(tensor->getName());
+}
+
+template<class T>
+T* getTensor(Blob* blob) {
+  assert(blob && "Blob is invalid");
+  if (blob && blob->template IsType<T>()) {
+    return blob->template GetMutable<T>();
+  }
+  return nullptr;
+}
+
+const caffe2::OperatorDef* getOpDef(const repr::NeuralNetOperator& nnOp) {
+  auto annotation = nnOp.getAnnotation();
+  if (!annotation || !isa<Caffe2Annotation>(annotation)) {
+    return nullptr;
+  }
+
+  return dyn_cast<Caffe2Annotation>(annotation)->getOperatorDef();
+}
+
+bool isOnIdeepDevice(const repr::NeuralNetOperator& nnOp) {
+  // We only want to fuse for IDEEP convs
+  const auto* op = getOpDef(nnOp);
+  if (op == nullptr || op->device_option().device_type() != DeviceType::IDEEP) {
+    return false;
+  }
+
+  return true;
+}
+
+bool shouldFuseConv(const repr::Conv& conv) {
+  return isOnIdeepDevice(conv) ? (conv.getGroup() <= 1) : false;
+}
+
+void resetConvForFusion(repr::NNGraph::NodeRef convNode, int fusion_type) {
+  // Fusion types:
+  // FUSION_CONV_RELU = 1
+  // FUSION_CONV_SUM = 2
+  // FUSION_CONV_SUM_RELU = 3
+  auto conv = repr::nn::get<repr::Conv>(convNode);
+  auto annotation = conv->getMutableAnnotation();
+  if (!annotation || !isa<Caffe2Annotation>(annotation)) {
+    return;
+  }
+
+  auto* op = dyn_cast<Caffe2Annotation>(annotation)->getMutableOperatorDef();
+  if (op == nullptr) {
+    return;
+  }
+
+  if (op->type() == "ConvFusion") {
+    assert(fusion_type == 1 && "Invalid nest fusion");
+    for (auto& arg : *op->mutable_arg()) {
+      if (arg.name() == "fusion_type") {
+        // Only from FUSION_CONV_SUM to FUSION_CONV_SUM_RELU
+        assert(arg.i() == 2 && "Invalid nest fusion");
+        arg.set_i(3);
+        return;
+      }
     }
-    const auto* op = dyn_cast<Caffe2Annotation>(annotation)->getOperatorDef();
+    return;
+  }
+
+  assert(fusion_type < 3 && "Invalid fusion type");
+  op->set_type("ConvFusion");
+  auto* arg = op->add_arg();
+  arg->set_name("fusion_type");
+  arg->set_i(fusion_type);
+}
 
-    // We only want to fuse for IDEEP convs
-    if (op->device_option().device_type() != DeviceType::IDEEP) {
-      return false;
+bool fuseConvBNHelperForIdeep(repr::NNModule* nn, caffe2::Workspace* ws) {
```

Comment:
`ConvBN` fusion isn't IDEEP specific. And I believe it's done in 
https://github.com/pytorch/pytorch/blob/master/caffe2/opt/fusion.cc#L17

Reply:
Yes. But difference tensor type and tensor API. It's difficult to invoke existing ConvBN fusion directly for Ideep tensors.